### PR TITLE
logWithNewLogMacros

### PR DESCRIPTION
**Pull request by HakonRydland**
## What type of PR is this.
1. [ ] Bug
2. [x] Change
3. [ ] Enhancement

### What have you changed and why?
Information: Moved away from A3A_fnc_Log in most files.
    

### Please specify which Issue this PR Resolves.
closes #XXXX

### Please verify the following and ensure all checks are completed.

1. [x] Have you loaded the mission in LAN host?
2. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [ ] No
2. [ ] Yes (Please provide further detail below.)

### How can the changes be tested?
Steps: 

********************************************************
Notes:
Not every log line has been tested obviously., but most should be fine, one thing that will cause issues is the use of ``JoinString`` inside of the macros, so if you find them scream at me to move the joinstring out of it.

Diag_logs are gonna be done separately along with few remaining A3A_fnc_log calls.

### CODE REVIEW

File: A3-Antistasi/functions/AI/fn_airbomb.sqf [hunk 1]
```diff
@@ -20,20 +20,20 @@ switch (_bombType) do
...
-	};
-	case ("NAPALM"):
+    };
+    case ("NAPALM"):
     {
-		_ammo = "ammo_Bomb_SDB";
+        _ammo = "ammo_Bomb_SDB";
         _bombOffset = 170;
-	};
-	default
```
Comment (wurzel0701) [inline A3-Antistasi/functions/AI/fn_airbomb.sqf:33]: Please dont do whitespace changes when changing other stuff, its hard to filter out which is which

**Review by wurzel0701 — APPROVED** [review]
Fucking hell @HakonRydland 20 minutes adventure my ass... :D

Changes are solid tho

**Comment by Bob-Murphy** [comment]
Weird invitation but ok

**Comment by wurzel0701** [comment]
Yeah, but an approval is an approval right

**Comment by HakonRydland** [comment]
hit the merge bofor th conflicts comes!!!!

File: A3-Antistasi/Includes/LogMacros.inc [hunk 2]
```diff
@@ -77,6 +106,10 @@ if (isNil "LogLevel") then {LogLevel = 2};
     #define ServerError_6(Message, A,B,C,D,E,F) if (0 < LogLevel) then { private _msg = format [Message,A,B,C,D,E,F]; LogServerMessage("Error", _msg) }
     #define ServerError_7(Message, A,B,C,D,E,F,G) if (0 < LogLevel) then { private _msg = format [Message,A,B,C,D,E,F,G]; LogServerMessage("Error", _msg) }
     #define ServerError_8(Message, A,B,C,D,E,F,G,H) if (0 < LogLevel) then { private _msg = format [Message,A,B,C,D,E,F,G,H]; LogServerMessage("Error", _msg) }
+    #define ErrorArray(Message, Array) if (0 < LogLevel) then { private _msg = [Message + " ["]; { _msg pushBack ("    " + str _x)} forEach Array; _msg pushBack "]"; _msg = _msg joinString endl; LogMessage("Error", _msg) }
+    #define ServerErrorArray(Message, Array) if (0 < LogLevel) then { private _msg = [Message +" ["]; { _msg pushBack ("    " + str _x)} forEach Array; _msg pushBack "]"; _msg = _msg joinString endl; LogServerMessage("Error", _msg) }
```
Comment (CalebSerafin) [inline A3-Antistasi/Includes/LogMacros.inc:110]: 1. This code can be cut down alot. Follow the same pattern for other Array logs.
```suggestion
    #define ServerErrorArray(Message, Array) if (0 < LogLevel) then { LogServerMessage("Error", (Message + " [" + endl + (Array apply {"    " + str _x} joinString endl) + endl + "]")) }
```

File: A3-Antistasi/Includes/LogMacros.inc [hunk 3]
```diff
@@ -77,6 +106,10 @@ if (isNil "LogLevel") then {LogLevel = 2};
...
+    #define ErrorArrayFormat(Message, Frmat, Array) if (0 < LogLevel) then { private _msg = [Message + " ["]; { _msg pushBack ( format (["    " + Frmat] + _x) ) } forEach Array; _msg pushBack "]"; _msg = _msg joinString endl; LogMessage("Error", _msg) }
+    #define ServerErrorArrayFormat(Message, Frmat, Array) if (0 < LogLevel) then { private _msg = [Message +" ["]; { _msg pushBack ( format (["    " + Frmat] + _x) ) } forEach Array; _msg pushBack "]"; _msg = _msg joinString endl; LogServerMessage("Error", _msg) }
```
Comment (CalebSerafin) [inline A3-Antistasi/Includes/LogMacros.inc:112]: 2. This code can be cut down alot. Follow the same pattern for other ArrayFormat logs.
```suggestion
    #define ServerErrorArrayFormat(Message, Frmat, Array) if (0 < LogLevel) then { LogServerMessage("Error", (Message + " [" + endl + (Array apply {format (["    " + Frmat] + _x)} joinString endl) + endl + "]")) }
```

**Review by CalebSerafin — CHANGES_REQUESTED** [review]
3. File lines need to be fixed* when using includes.  *NOTE the line numbers will be 1 less, I have not found a way around that without manual entering.
Add something like this snippet to common or somewhere else
```h
// https://github.com/Sparker95/Vindicta/blob/224980f20c9bd43c3c1a2dd2f183b4fa5d5975b9/src/common.h#L13-L18
// Some BS to force Arma preprocessor line numbers to be correct.
// You basically have to put this after every pre-processor block if you want the correct line numbers
#define FIX_LINE_NUMBERS2(sharp) sharp##line __LINE__ __FILE__
// Use this with () but WITHOUT terminating ; like:
// FIX_LINE_NUMBERS()
#define FIX_LINE_NUMBERS() FIX_LINE_NUMBERS2(#)
```

When using include:
```c
#include "..\..\Includes\LogMacros.inc"
FIX_LINE_NUMBERS()
```

4. In the header of LogMacros explicitly state that multi-element arrays cannot be defined within the macro: ie: `MACRO([1,2,3])` is split into macro arguments.

5. For viewing pleasure with some syntax highlighters.
https://github.com/official-antistasi-community/A3-Antistasi/blob/477014f9672d606dc05a9abcabe5ed077be6b175/A3-Antistasi/Includes/LogMacros.inc#L79
Could be modified to use single quotes:
```hpp
private _index = _fileName find '\'; 
```

6. Missed this earlier:
For remoteExec code to server: A simple function such as 
```sqf
diag_log text _this;
```
will be required to avoid sending structured text over the network. (Inefficient + log spam)
Usage:
```sqf
    #define LogServerMessage(Level, Message) ((systemTimeUTC call _LOG_timeText) + " | " + _LOG_prefix + " | "+ Level +" | File: " + (__FILE__ call _LOG_file) + " | " + Message + format [" | Client: %1 [%2]", player, clientOwner]) remoteExec ["LOCAL_LOG_FUNCTION", 2]
```

7. Minor, code could be pulled out of the include into functions or preinit function. It will not need to be redefined multiple times or check existance of log level.
```sqf
//filename parser
private _LOG_file = {
    private _fileName = reverse _this;
    private _index = _fileName find '\';
    reverse (_fileName select [0,_index]);
};
private _LOG_timeText = {str (_this#0) +"-"+ str (_this#1) +"-"+ str (_this#2) +" "+ str (_this#3) +":"+ str (_this#4) +":"+ str (_this#5) +":"+ str (_this#6)};
if (isNil "LogLevel") then {LogLevel = 2};
```

8. Minor, The `_LOG_file` function calls can be replaced by a one-liner that runs 35% faster than the larger function.
```sqf
(__FILE__ select [count __FILE__ -(reverse __FILE__ find '\'),1e6])
// ⬇
#define LogMessage(Level, Message) diag_log text ((systemTimeUTC call _LOG_timeText) + " | " + _LOG_prefix + " | "+ Level +" | File: " + (__FILE__ select [count __FILE__ -(reverse __FILE__ find '\'),1e6]) + " | " + Message)
```

NB: Remember to test after changes were made.

**Comment by CalebSerafin** [comment]
Note, All logs that have commas: define the string somewhere else; replace commas with semi-colons or other grammatical replacements.

**Review by CalebSerafin — APPROVED** [review]
Solid stuff. All my concerns have been addressed. The roll-out should go unhitched.

**Pull request merged**